### ISSUE TITLE
Feature Request: Improve Chinese semantic search quality

### ISSUE THREAD

**Issue by JerryLiu720** [milla-jovovich/mempalace#516]
## Problem

Semantic search has very poor quality for Chinese queries. Chinese content cannot be found even when it exists in the palace.

### Example 1: Document exists but cannot be found

I have a document about deployment procedures (部署流程文档) in my palace. Chinese query returns no results or very low match scores, but English query and exact text search both work fine. The document exists and contains "部署流程", but semantic search cannot find it with Chinese keywords.

### Example 2: Negative match scores for Chinese queries

When I search for "工作总结" (work summary), I get Match: -0.016 with completely irrelevant results. This is for documents that actually contain "工作总结" in the content.

### Example 3: Same meaning, different results

Searching for "技术文档" (technical documentation) in Chinese gives Match scores below 0.2 with poor results, while searching for "technical documentation" in English gives Match scores above 0.8 with good results. Both mean the same thing, but Chinese gets dramatically worse results.

## Impact

This makes semantic search nearly unusable for Chinese users. Cannot find documents by Chinese titles or content, have to remember English translations, must fall back to exact text search using rg, which defeats the purpose of semantic search.

### Current workaround

1. Use English keywords instead of Chinese
2. Use technical terms (table names, function names)
3. Use rg for exact search: rg "关键词" ~/myspace
4. Ask AI to search (AI uses knowledge graph, not semantic search)

## Expected Behavior

Chinese queries should return semantically relevant results, just like English queries do.

## Environment

- OS: macOS
- mempalace version: 3.1.0
- Language: Chinese (Simplified)

## Additional Context

The benchmark code already supports multiple embedding models, suggesting the architecture can support this. The current default model works great for English but appears to have limited Chinese language support. This likely affects Japanese, Korean, and other non-English users as well. Thank you for this excellent project!

**Comment by web3guru888** [comment]
This is a real problem — and you have already diagnosed it correctly. The default `all-MiniLM-L6-v2` model was trained primarily on English text. Its Chinese vocabulary coverage is poor, so Chinese queries produce embeddings that are essentially noise relative to the document vectors, which is why you see match scores near zero or negative.

The fix is a multilingual embedding model. PR #488 by EndeavorYen goes directly at this — it adds `paraphrase-multilingual-MiniLM-L12-v2` as an optional embed model, which supports 50+ languages including Chinese. If you can test against that branch on your actual palace, that would be extremely valuable feedback since your use case (mixed Chinese/English content, semantic search across both) is exactly what it is targeting.

A few things worth knowing:

**Model compatibility**: changing the embedding model on an existing palace requires a full re-mine. The vector space changes completely, so old vectors from `all-MiniLM-L6-v2` and new vectors from the multilingual model cannot be mixed. A `--rebuild` flag would be needed.

**Cross-lingual search**: `paraphrase-multilingual-MiniLM-L12-v2` is trained for cross-lingual alignment, meaning a Chinese query for "工作总结" should surface English documents about "work summaries" and vice versa. This is the behavior you want.

**Performance trade-off**: the multilingual model is slightly larger (~135MB vs ~80MB), which is minor but worth noting for first-run.

In the meantime, the workarounds you listed (function names, table names, rg fallback) are reasonable — exact terms are embedding-model-agnostic since they land in the same vector neighbourhood regardless of language. Your AI using the knowledge graph is actually hitting the right layer — the KG entities and triples are text, not vectors, so they are not affected by this.

---
*[MemPalace-AGI integration — production stats at https://milla-jovovich.github.io/mempalace/integrations/mempalace-agi/]*

**Comment by JerryLiu720** [comment]
@web3guru888 Thank you for the detailed explanation and for pointing me to PR #488! 

I'm very interested in testing it with my actual palace. My use case is exactly what #488 targets: mixed Chinese/English technical documentation, SQL scripts with Chinese comments, and work notes primarily in Chinese.

I see #488 is already in good shape (654 tests passing, LGTM from your side). A few questions before I start testing:

**1. Testing setup**

Should I test like this?

```bash
git clone https://github.com/milla-jovovich/mempalace.git
cd mempalace
git checkout feat/multilingual-support
pip install -e .[multilingual]
```

**2. Re-mining my palace**

You mentioned `--rebuild` flag for re-mining. Is this the correct workflow?

```bash
# Backup first
cp -r ~/.mempalace ~/.mempalace.backup

# Re-mine with multilingual model
mempalace mine ~/myspace --rebuild
```

**3. Testing scope**

My palace has ~500 files. I'll test:
- Chinese semantic search quality (the main issue)
- Cross-lingual search (Chinese query → English docs)
- Mixed-language documents
- Performance comparison with current version

**4. Feedback format**

Should I report back in #488 or here in #516? I can provide:
- Before/after match scores for Chinese queries
- Specific examples where multilingual model improves results
- Any edge cases or issues I encounter

Let me know if this testing plan makes sense!

**Comment by web3guru888** [comment]
@JerryLiu720 That testing plan is solid — happy to clarify each step before you run it.

**1. Testing setup**

Your checkout is correct, but add the model cache step:

```bash
git clone https://github.com/milla-jovovich/mempalace.git
cd mempalace
git checkout feat/multilingual-support
pip install -e .[multilingual]

# Pre-download the multilingual model before re-mining
# (avoids timeout mid-run on large palaces)
python -c "from sentence_transformers import SentenceTransformer; SentenceTransformer('paraphrase-multilingual-mpnet-base-v2')"
```

The model is ~1 GB. Downloading it explicitly first means re-mining doesn't hang silently if your network is slow.

**2. Re-mining workflow**

Your workflow is correct. One addition: back up the ChromaDB state too, not just the file tree:

```bash
cp -r ~/.mempalace ~/.mempalace.backup            # file metadata + KG
cp -r ~/.chroma ~/.chroma.backup 2>/dev/null || true  # vector store (if default location)
mempalace mine ~/myspace --rebuild
```

The `--rebuild` flag re-mines everything from scratch using the new model. With 500 files this should take a few minutes, not hours. The KG (if you have one) is model-independent — it persists across re-mines.

**3. Testing scope**

Good scope. One addition worth tracking: **same-query score drift** — pick 5–10 queries you use regularly and record the top-3 results + similarity scores before and after. The multilingual model will shift scores for English-only content too (usually slightly lower for short English queries), so you want to know the trade-off, not just the win.

For cross-lingual: try both directions — Chinese query against English docs, and English query against Chinese docs. PR #488's model handles both.

**4. Feedback format**

Report in **PR #488** — that's where EndeavorYen is tracking test results and the before/after scores you collect will be exactly what they need for the review. Tag the issue number (`#516`) in your comment so there's a cross-link. If you hit any `--rebuild` failures specific to your setup, opening a separate issue is fine too.

Good luck with the 500-file palace — that's a good real-world test size.

**Comment by fuzzymoomoo** [comment]
Wholly unrelated but I find this fascinating, I recently did work on mining data out of conversation transcripts and things got real interesting when  I started working with the Zulu language, there are some nice parallels between Zulu and Chinese in the way information is conveyed vs English.

**Comment by web3guru888** [comment]
@fuzzymoomoo — that's a genuinely interesting parallel. I'd never thought about Zulu in this context but you're right: both Zulu and Chinese encode a lot of meaning in morphological structure (Zulu's noun class agreement, Chinese's classifier system) that English distributes across separate words. An embedding model trained primarily on English is going to compress those relationships differently, probably losing precision on both ends.

Did you find that the multilingual models handled Zulu noticeably better, or was it still hit-or-miss? I'm curious whether the issue is at the tokenization layer (BPE treating Zulu agglutinative forms as noise) or at the semantic layer. @JerryLiu720 might find the comparison useful — similar root problem even if the failure modes are different.

**Comment by fuzzymoomoo** [comment]
@web3guru888 I never even got to testing it out on LLM, the first step was trying to find areas of the transcript that should be looked at by the LLM to do the data mining using regex, I was taking the route of using a scripted text, had it translated into Zulu, but I noticed when I took that and translated back to English there was already a loss or difference in/of meaning, subtle but there, this didn't bode well for a good result and I moved onto other things, I was running a local LLM on a tiny GPU so I never tested multilingual models, but that's what i find interesting here. I guess the issue in my case was semantic, trying to build regex expressions for a phrase in a language I know little about and using English as guidance, lots of information was "lost" so I'm curious how this will play out in mempalace or even with the things I have been playing around with, could you do a CDD handoff to another actor speaking an entirely different language (at the end of the day these differences also shape how we think and perceive information, be it AI or Human)

**Comment by web3guru888** [comment]
The round-trip translation experiment you ran is actually a really elegant way to measure semantic fidelity — you weren't just testing Zulu, you were testing whether the meaning survived the round trip, and the subtle differences you noticed *are* the loss function made visible. Regex on the Zulu output was always going to be fighting a corrupted reference, because the ground truth you built the patterns against (the English original) wasn't what the model actually stored.

The CDD-across-languages question is one I haven't seen raised seriously here before, and it's genuinely interesting. In the CDD framing, the handoff is a structured memory transfer — context, state, open threads. If the receiving actor works in a different language, the question is: where does the semantic content live, and does translation happen at the right layer?

What MemPalace's architecture actually gives you here is potentially more robust than an LLM-level handoff: the palace stores verbatim text (so the original language is preserved, not homogenized into English-centric embeddings), and KG triples are essentially language-agnostic labels once they're structured. An actor working in Spanish or Zulu could retrieve from the same palace if they're using a multilingual embedding model — the cross-lingual bridge is at retrieval time, not at write time. That's a fundamentally different approach from "translate the handoff document and hope nothing gets lost."

Your point about language shaping perception hits something real though. Embedding models don't just encode language — they encode the cultural semantic space their training corpus came from. A model trained predominantly on English text has learned that certain concepts cluster together in ways that might not match how a Zulu or Mandarin speaker would organize the same ideas. The embeddings carry that bias quietly. For multilingual CDD, you'd want to think carefully about whether your retrieval model's latent space actually bridges those cultural concept clusters, or just does surface-level translation.

The practical ideal is probably: store verbatim in the language of origin, use a multilingual model (mE5, LaBSE, etc.) for embedding, and let each actor retrieve in their own language without ever having to "translate the palace." Whether current MemPalace + any given multilingual model actually delivers that fidelity in practice — that's the experiment I'd want to run.

**Comment by vincent067** [comment]
👋 大家好！我是一位来自深圳的开发者，目前在做中文场景下的 AI Agent 系统，所以对这个问题感触很深。\n\n经过分析和实际测试，默认的 `all-MiniLM-L6-v2` 确实在中文语料上表现不佳——embedding 空间几乎是"噪音级"的，导致语义搜索完全失效。这不仅是用户体验问题，对中文开发者来说几乎是 adoption blocker。\n\n我非常支持 #553 和 #488 的方向：**把 embedding model 做成可配置项**。从我们的使用经验来看，`intfloat/multilingual-e5-base`（512 token 窗口，跨语种对齐更好）可能比 `paraphrase-multilingual-MiniLM-L12-v2` 更适合长文本技术文档。当然，最终还是要看 benchmark 数据说话。\n\n如果维护者愿意推进这个特性，我很乐意参与测试中文场景，或者补充一份中文 README 的 setup 示例。\n\n感谢维护者的辛苦工作！期待看到 MemPalace 在中文环境里也能发挥全部潜力 🙏

**Comment by web3guru888** [comment]
@vincent067 — welcome, and great context from someone actually running Chinese AI agent pipelines in production.

The model recommendation is interesting — `intfloat/multilingual-e5-base` is a reasonable alternative to `paraphrase-multilingual-MiniLM-L12-v2`. The key differences worth benchmarking:

- **Token window**: multilingual-e5-base has a 512-token limit vs 128 for paraphrase-multilingual-MiniLM-L12-v2. For technical docs and conversation transcripts, this matters — 128 tokens cuts off mid-sentence in dense Chinese text.
- **Training objective**: e5-base uses a contrastive text-pair approach that tends to give better cross-lingual alignment on asymmetric query/document pairs, which is exactly what `mempalace search` does (short query against longer stored memories).
- **Size tradeoff**: e5-base is heavier (~270M parameters vs ~118M), which may be a concern for users running on lower-spec machines.

Our testing with #488's approach (paraphrase-multilingual-MiniLM-L12-v2) showed a large improvement over the default for Chinese, but we haven't directly benchmarked e5-base on the same palace. If you run that comparison, the data would be really valuable for the #553 decision — specifically hit rate on Chinese queries, and whether the 512-token window helps on longer stored entries.

The Chinese README setup example is a great idea. Even a minimal block showing `MEMPALACE_MODEL=intfloat/multilingual-e5-base` with the caching path would help Chinese developers skip the "why doesn't this work" phase entirely.

**Comment by NickShtefan** [comment]
Chiming in with concrete numbers from the Russian side: same root cause, same fix. On a 15K-drawer palace, switching from default `all-MiniLM-L6-v2` to `intfloat/multilingual-e5-base` lifted average similarity from 0.19–0.40 (essentially noise) to 0.70–0.77 on production queries. PR #442 makes that switch a one-flag operation: `mempalace init --model intfloat/multilingual-e5-base`, model gets stamped into ChromaDB collection metadata, MCP and CLI both read from there so they can't desync.

#442 also closes the silent `_get_collection` bug (#903) where MCP search would silently fall back to 384-dim and return zero hits against a 768-dim palace — that one bites Chinese palaces just as hard as Russian ones. Just rebased onto `develop` today, ready for review. If anyone here wants to validate against a Chinese corpus before merge, fork's at `NickShtefan/mempalace#feat/multilingual-embedding`.
